### PR TITLE
chore(PE-3569): standardize ant source

### PR DESCRIPTION
**Pull request by atticusofsparta**

### CODE REVIEW

File: coverage/clover.xml
Comment (kunstmusik) [inline coverage/clover.xml:1]: Remove coverage folder from commit and add to .gitignore as it is generated at test-time.

File: contracts/ANT-1/tests/evolve.test.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import Arweave from "arweave";
+import fs from "fs";
+import path from "path";
+import { ANTState } from "../types";
+import { ANTDeployer } from "../utils";
+
+describe("Testing evolve...", () => {
+  const _arweave: Arweave = global.arweave;
+  const _wallets: Record<string, JWKInterface> = global.wallets;
+  const _warp: Warp = global.warp;
```
Comment (dtfiedler) [inline contracts/ANT-1/tests/evolve.test.ts:12]: no need to do `_` prefixes

File: coverage/clover.xml
Comment (dtfiedler) [reply coverage/clover.xml:1]: +1

File: contracts/ANT-1/actions/write/setName.ts [hunk 2]
```diff
@@ -23,10 +24,10 @@ export const setName = async (
...
   // check name validity
   if (
-    typeof name !== "string" || // must be a string
-    name === ""
+    typeof name !== 'string' || // must be a string
+    name === ''
   ) {
-    throw new ContractError("Invalid ANT name");
+    throw new ContractError('Invalid ANT name');
   }
```
Comment (dtfiedler) [inline contracts/ANT-1/actions/write/setName.ts:31]: use AJV

File: contracts/ANT-1/actions/write/setRecord.ts [hunk 3]
```diff
@@ -23,32 +29,32 @@ export const setRecord = async (
...
+    typeof subDomain !== 'string' || // must be a string
     subDomain.length > MAX_NAME_LENGTH || // the name is too long
-    (!nameRes && subDomain !== "@") || // the name does not match our regular expression and is not the root
-    subDomain === "www" // this is a reserved name
+    (!nameRes && subDomain !== '@') || // the name does not match our regular expression and is not the root
+    subDomain === 'www' // this is a reserved name
   ) {
-    throw new ContractError("Invalid ArNS Record Subdomain");
+    throw new ContractError('Invalid ArNS Record Subdomain');
   }
```
Comment (dtfiedler) [inline contracts/ANT-1/actions/write/setRecord.ts:42]: use AJV

File: contracts/ANT-1/contract.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import { removeController } from './actions/write/removeController';
+import { removeRecord } from './actions/write/removeRecord';
+import { setController } from './actions/write/setController';
+import { setName } from './actions/write/setName';
+import { setRecord } from './actions/write/setRecord';
+import { setTicker } from './actions/write/setTicker';
+import { transferTokens } from './actions/write/transferTokens';
+import { ANTState, ContractResult, PstAction } from './types';
+
+declare const ContractError;
```
Comment (dtfiedler) [inline contracts/ANT-1/contract.ts:14]: create `globals.d.ts` and add `ContractError` and `SmartWeave` similar to https://github.com/ar-io/testnet-contract/blob/develop/src/globals.d.ts and remove the declarations in each handler function

File: contracts/ANT-1/tests/evolve.test.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import { ANTState } from '../types';
+import { ANTDeployer } from '../utils';
+
+describe('Testing evolve...', () => {
+  const arweave: Arweave = global.arweave;
+  const wallets: Record<string, JWKInterface> = global.wallets;
+  const warp: Warp = global.warp;
+  const defaultOwner = Object.entries(wallets)[0];
+
+  let _srcTxId: string;
```
Comment (dtfiedler) [inline contracts/ANT-1/tests/evolve.test.ts:16]:
```suggestion
  let _srcTxId: string;
```
```suggestion
  let srcTxId: string;
```

File: contracts/ANT-1/tests/evolve.test.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+      'utf8',
+    );
+    const srcTx = await warp.createSource(
+      { src: contractSource },
+      Object.values(wallets)[0],
+      true,
+    );
+    _srcTxId = await warp.saveSource(srcTx, true);
+    await mineBlock(arweave);
+  });
```
Comment (dtfiedler) [inline contracts/ANT-1/tests/evolve.test.ts:30]: does this need to be done `beforeEach`? it's a lot of overhead for each test to create a new source. you can likely just create one in `setup.jest.ts` and use the ID for testing interactions

File: contracts/ANT-1/tests/removeController.test.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  const wallets: Record<string, JWKInterface> = global.wallets;
+  const warp: Warp = global.warp;
+  const defaultOwner = Object.entries(wallets)[0];
+
+  it('Should remove controller from the ANT', async () => {
+    const ANT = await ANTDeployer(warp, {
+      address: defaultOwner[0],
+      wallet: defaultOwner[1],
+    });
+    await mineBlock(arweave);
```
Comment (dtfiedler) [inline contracts/ANT-1/tests/removeController.test.ts:19]:
```suggestion
```

File: contracts/ANT-1/tests/removeController.test.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    const contract = warp.contract<ANTState>(ANT).connect(defaultOwner[1]);
+    const result = await contract.writeInteraction({
+      function: 'removeController',
+      target: defaultOwner[0],
+    });
+
+    expect(result).toBeDefined();
+    expect(result?.originalTxId).toBeDefined();
+
+    await mineBlock(arweave);
```
Comment (dtfiedler) [inline contracts/ANT-1/tests/removeController.test.ts:30]:
```suggestion
```
all `await mineBlock(arweave)` calls can be removed

File: .github/build_and_test.yml [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+          cache: 'yarn'
+      - run: yarn install
+      - run: yarn ${{ matrix.step }}
+        env:
+          ARNS_CONTRACT_TX_ID: ${{ vars.ARNS_CONTRACT_TX_ID }}
+      - name: Upload to Codecov
+        if: matrix.step == 'test:unit'
+        uses: codecov/codecov-action@v3
+        env:
+          CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
```
Comment (dtfiedler) [inline .github/build_and_test.yml:33]: comment out for now

File: .husky/pre-push [hunk 10]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env sh
+. "$(dirname -- "$0")/_/husky.sh"
+. "$(dirname -- "$0")/common.sh"
+
+yarn run build
+yarn run test:integration
```
Comment (dtfiedler) [inline .husky/pre-push:6]: let's not have a pre-push for now, can remove this file

File: jest.config.js [hunk 11]
```diff
@@ -8,6 +24,7 @@ module.exports = {
   testEnvironment: 'node',
 
   transformIgnorePatterns: ['<rootDir>/node_modules/(?!@assemblyscript/.*)'],
+  testTimeout: 600_000,
```
Comment (dtfiedler) [inline jest.config.js:27]: to high - set it to 30_000

File: .github/build_and_test.yml [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    permissions:
+      actions: write
+      id-token: write
+      contents: read
+      checks: write
+      packages: write
+      pull-requests: write
+    strategy:
+      matrix:
+        step: ['format:check', 'lint:check', 'build', 'test:unit']
```
[the remark this answers is not in the document]
Comment (dtfiedler) [reply .github/build_and_test.yml:18]: ah i see you added the test:unit command. that's fine, can leave. but add `test:integration` here as well

File: src/actions/write/setRecord.ts [hunk 13]
```diff
@@ -23,32 +45,32 @@ export const setRecord = async (
   }
 
   // check subdomain validity
-  const namePattern = new RegExp("^[a-zA-Z0-9_-]+$"); // include underscores and dashes
+  const namePattern = new RegExp('^[a-zA-Z0-9_-]+$'); // include underscores and dashes
```
Comment (dtfiedler) [inline src/actions/write/setRecord.ts:48]: add AJV validations to these methods similar to what we do in testnet contract and remove these input checks

File: src/types.ts [hunk 14]
```diff
@@ -1,22 +1,39 @@
...
-      transactionId: string, // The transaction ID that the subdomain points to.
-      ttlSeconds: number // The amount of time (in seconds) this transaction is cached for, default at 900 seconds
-    }
+  records: {
+    // A list of all subdomains and their corresponding Arweave Transaction IDs and TTLs
+    [subDomain: string]: {
+      // the subdomain used, default is the root @
+      transactionId: string; // The transaction ID that the subdomain points to.
+      ttlSeconds: number; // The amount of time (in seconds) this transaction is cached for, default at 900 seconds
+    };
```
Comment (dtfiedler) [inline src/types.ts:29]: create types:

```typescript
export type ArNSRecordData = {
   transactionId: string,
   ttlSeconds: number,
}

export type ArNSRecords = Record<string, ArNSRecordData>;
```

and set the type here
```
    records: ArNSRecords
```

File: .github/build_and_test.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (atticusofsparta) [reply .github/build_and_test.yml:18]: Yup - and correct, we dont have unit tests (yet, but soon)

File: .github/build_and_test.yml (hunk 9 above)
Comment (dtfiedler) [inline .github/build_and_test.yml:33]: token has been added

File: .github/deploy.yml [hunk 15]
```diff
@@ -0,0 +1,37 @@
+name: Build / Test / Deploy
+
+on:
+  workflow_dispatch:
+  push:
+    branches:
+      - develop
+      - main
```
Comment (dtfiedler) [inline .github/deploy.yml:8]: let's disable this on push for now (comment this out) - until we validate via manual invocations it is doing what we expect

Comment (atticusofsparta) [reply .github/deploy.yml]: removed

File: .github/build_and_test.yml (hunk 9 above)
Comment (atticusofsparta) [reply .github/build_and_test.yml]: Should I re-add this with the pill on the read me?

File: src/types.ts (hunk 14 above)
Comment (atticusofsparta) [reply src/types.ts]: added

File: src/README.md [hunk 16]
```diff
@@ -0,0 +1,217 @@
+## Table of Contents
```
[the remark this answers is not in the document]
Comment (atticusofsparta) [reply src/README.md]: moved

File: jest.config.js (hunk 11 above)
Comment (atticusofsparta) [reply jest.config.js]: lowered - it works

File: .husky/pre-push (hunk 10 above)
Comment (atticusofsparta) [reply .husky/pre-push]: removed

File: .github/build_and_test.yml (hunk 9 above)
Comment (dtfiedler) [reply .github/build_and_test.yml:33]: Don't worry about the pill

File: README.md [hunk 17]
```diff
@@ -1,9 +1,16 @@
+[![codecov](https://codecov.io/gh/ar-io/ant-pilot/graph/badge.svg)](https://codecov.io/gh/ar-io/ant-pilot)
```
Comment (dtfiedler) [inline README.md:1]: delete for now

File: README.md [hunk 18]
```diff
@@ -1,9 +1,16 @@
...
 # Arweave Name Token Proof of Concept
 
 ## Project setup
```
Comment (dtfiedler) [inline README.md:4]: Move this down to developer section at the bottom

File: README.md [hunk 19]
```diff
@@ -19,33 +26,239 @@ yarn build
...
+
+#### Schemas
+
+We have defined multiple schemas for different contract methods, such as `balanceSchema`, `setControllerSchema`, `setNameSchema`, etc. These schemas are located in the `./schemas` directory. Each schema precisely defines the expected format of inputs for a corresponding contract method.
+
+#### Compilation of Validation Functions
+
+In the build process, we create an instance of AJV and add our schemas to it. We then use `standaloneCode` from `ajv/dist/standalone` to compile these schemas into standalone validation functions. This compilation process generates a module (`validations.js`) in the `src` directory, which contains all the validation functions.
+
+#### Usage in Contract Methods
```
Comment (dtfiedler) [inline README.md:208]: Don't think it's necessary to get into the nitty gritty of AJV

File: README.md [hunk 20]
```diff
@@ -19,33 +26,239 @@ yarn build
...
+}
+```
+
+# Additional Resources
+
+### Documentation
+
+- [Arweave Name System Documentation]
+- [Permaweb Cookbook]
+- [AJV]
```
Comment (dtfiedler) [inline README.md:228]: Add links to portal app and arns service repo

**Review by dtfiedler — COMMENTED** [review]
A few small README suggestions

File: README.md (hunk 20 above)
Comment (atticusofsparta) [reply README.md:214]: added

File: README.md (hunk 19 above)
Comment (atticusofsparta) [reply README.md]: abbreviated

File: README.md (hunk 18 above)
Comment (atticusofsparta) [reply README.md]: moved

File: README.md (hunk 17 above)
Comment (atticusofsparta) [reply README.md]: removed

**Pull request merged**